File: java/src/DynamicProgrammingSegmentation.java
/*
 * This code performs the main segmentation task by a dynamic programming approach, using the transformed version of trajectories.
 */

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

import DataStructures.MeanStd;

public class DynamicProgrammingSegmentation {

	public static void main(String[] args) throws IOException {

		List<String> probDissData = Files.readAllLines(Paths.get("prerequisiteFiles/ProbabilisticDissimilarities.csv"));
		BufferedWriter bw = new BufferedWriter(new FileWriter("output/segmentation_results.csv"));
		bw.write("TripId,TimeStep,Speed,Acceleration,HeadingChange,Latitude,Longitude,PMD,StartOfSegment\n");
		
		// You can modify the maximum number of segments for a trajectory, which is an input to MDL process.
		// The choice of 50 is based on this observation that we usually observe less than this number of segments to be found for a trajectory of moderate size (between 5 to 20 minutes)
		int maximum_number_of_segments = 50;
		
		int n_trips = 0;
		
		String crnt_trip = "";
		List<Double> points = new ArrayList<Double>();
		List<String> trip_points = new ArrayList<>();
		
		// read probabilistic dissimilarity files line by line, separate trajectories, perform segmentation, and print out the results
		for(int i=1; i < probDissData.size(); i++){
			String[] parts = probDissData.get(i).split(","); /* TripId,TimeStep,ProbDissimilarity,Lat,Lng,Speed,Acceleration,Heading */
			if(parts[0].equals(crnt_trip)){
				points.add(Double.parseDouble(parts[2]));
				trip_points.add(parts[0] + "," + parts[1] + "," + parts[5] + "," + parts[6] + 
						"," + parts[7] + "," + parts[3] + "," + parts[4] + "," + parts[2] + ",");
			}
			else{
				if (crnt_trip.length() > 0){
					// do the segmentation
					System.out.print("Segmenting trajectory " + crnt_trip + " ");
					mainSegmentationProcess(points, trip_points, bw, maximum_number_of_segments);
					System.out.println(" done!");
					n_trips += 1;
				}
				crnt_trip = parts[0];
				points = new ArrayList<Double>();
				trip_points = new ArrayList<>();
				points.add(Double.parseDouble(parts[2]));
				trip_points.add(parts[0] + "," + parts[1] + "," + parts[5] + "," + parts[6] + 
						"," + parts[7] + "," + parts[3] + "," + parts[4] + "," + parts[2] + ",");
			}
		}
		
		//do segmentation for the last trajectory
		System.out.print("Segmenting trajectory " + crnt_trip + " ");
		mainSegmentationProcess(points, trip_points, bw, maximum_number_of_segments);
		System.out.println(" done!");
		n_trips += 1;
		
		bw.close();		
		System.out.println("\nThe segmentation process is finished for " + n_trips + " trips!");
	}
	
	public static MeanStd estimateGuassian(List<Double> input, int start, int end){
		MeanStd ms = new MeanStd();
		ms.mean = 0;
		ms.std = 0;
		
		int size = end - start + 1;
		
		for(int i=start; i <= end ; i++)
			ms.mean += input.get(i);		
		ms.mean /= size;
		
		for(int i=start; i <= end ; i++)
			ms.std += Math.pow(input.get(i) - ms.mean, 2);
		
		ms.std = Math.sqrt((1.0/size) * ms.std);	
		 
		return ms;
	}
	
	public static double lnOfNormalDistribution(double x, double mean, double std){
		double value = 0;		
		//value = ln(N(x, mean, std)) where N stands for normal distribution. N(x, mean, std) = (1/std * (2PI)^0.5) * exp(- (x - mean)^2 / (2 * std^2) )
		value = -1 * ((Math.log(std * Math.sqrt(2 * Math.PI))/Math.log(Math.E)) + (Math.pow(x-mean, 2)/(2 * Math.pow(std, 2))));	
		return value;
	}
	
	public static double[][] calculateDelta(List<Double> points){
		//** Calculation of Delta for all form of segments in given trajectory
		//delta_i(n_(i-1) , n_i - 1): the formula before (3) in paper! where n_(i-1) and n_i - 1 can be any possible pairs in trajectory!
		//Simply, here we have a window of dynamic size which scans whole trajectory! we will find the likelihood of that windows based on... 
		//...distribution of its points. Such distribution is a normal distribution and we have the assumption of independence of nodes ...
		//... in order to make simplification in problem
		
		double[][] delta = new double[points.size() - 1][points.size()];
		
		for(int i=0; i < delta.length; i++){
			for(int j=0; j < delta[i].length; j++){
				delta[i][j] = Double.MAX_VALUE; //some form of initialization
				if((j - i + 1) >= 2){//the length of a segment should be at least 2!					
					//get the distribution
					MeanStd ms = estimateGuassian(points, i, j);		
					//get the ln value based on points in this segment and their corresponding distribution
					double value = 0;
					//The goal is minimizing the delta! So, when we have standard deviation = 0, this means no changes is observing in ...
					//...probabilistic dissimilarity values in range i to j! So, no change, no std! In other words, we have a uniform distribution...
					//...for this range! In this way, I decided to let the delta to be 0 in such situation. 
					if(ms.std != 0){
						for(int k=i; k <= j; k++)
							value += lnOfNormalDistribution(points.get(k), ms.mean, ms.std);
						//We have a negative sign behind the ln in original formulation
						value *= -1.0;
					}
					delta[i][j] = value;
				}
			}
		}
		return delta;
	}
	
	public static void dynamicProgramingSegmentation(List<Double> points, int Ns, double[][] delta, double[][] I, int[][] Index){		
		//** Finding optimal segmentation by having N_s as number of existing segments in given trajectory
		//Here, we will find I_k(L) for all k = 1,2,...,N_s and for all L = 1,2,...,N
		//The approach for finding these values is based on dynamic programming formula...
		//... which is provided as relation (4) in paper
		
		//initialization of I
		for(int i=0; i < I.length; i++)
			for(int j =0; j < I[i].length; j++)
				I[i][j] = Double.MAX_VALUE;
		
		//In addition to minimum values, we need minimum indexes which show the best breaking points for segments		
		//initialization of Index
		for(int i=0; i < Index.length; i++)
			for(int j =0; j < Index[i].length; j++)
				Index[i][j] = 0;
		
		//now, I am going to find the optimal values
		for(int k=0; k < Ns; k++)
			
			//Base case: this mean just having a single segment or I_1(L)
			//Note: here k is 0 but based on paper formulation is 1
			if(k == 0){
				//here, the assumption is L>=2; but I considered this once that delta is calculated
				for(int L=0; L < points.size(); L++)
					I[k][L] = delta[k][L]; //look, segments of length 0 or 1 have delta value as +infinity!
			}
		
			//Rest of the cases (having more than one segment): Well, here is the idea of dynamic programming
			else{
				//why (k*2 + 1): look at the printed picture! I have to have some lower level for L! when I have a couple of segments, ... 
				//...the minimum criteria is I have a segment from 0 to 1. And the 2nd segment would be from 2! So, when k=1, the min length...
				//... for L should be 3 in order to have at least two separate segments. 			
				for(int L=(k*2 + 1); L < points.size(); L++){
					double minValue = Double.MAX_VALUE;
					int minIndex = -1;
					//I. Why (L-1)? since I want to let the last segment be at least from L-1 to L. That is, the minimum length is 2!
					//II. Look, here I have equality condition for L not just lower than! since L is index of upper loop and is precise. 
					//III. Why nk_1 = k*2? look at the last line of formula in left column of page 2 of paper. we have nk_1 as starting index ...
					//... of last segment. When have k = 1 or just want two segments, such index is at least 2! this means k*2. 
					for(int nk_1 = k*2; nk_1 <= L-1; nk_1++){
						double value = I[k-1][nk_1 - 1] + delta[nk_1][L];
						if(value < minValue){
							minValue = value;
							minIndex = nk_1;
						}
					}
					I[k][L]     = minValue;
					Index[k][L] = minIndex;
				}				
			}
	}
	
	public static double MinimumDescriptionLength(List<Double> points, int[][] Index, int Ns){
		//Here, the goal is to find the goodness of fit without any kind of over-fitting or a problem which have unseen data. 
		//We use the formula (7) in paper which is written based on a well known research paper: "modeling by shortest data description, 1978"		
		double MDL = 0;
		
		//1. Calculation of first term
		double mle = 0;

		int segmentEnd   = points.size()-1;
		int segmentBegin = -1;
		for(int k=Ns-1; k>= 0; k--){
			double thisSegmentMLE = 0;
			segmentBegin = Index[k][segmentEnd];
			MeanStd ms = estimateGuassian(points, segmentBegin, segmentEnd);
			for(int i=segmentBegin; i < segmentEnd; i++)
				thisSegmentMLE += lnOfNormalDistribution(points.get(i), ms.mean, ms.std);
			mle += thisSegmentMLE;
			
			segmentEnd = segmentBegin -1;
		}
		//there is a negative sign behind the ln in (7)
		mle *= -1;
		MDL = mle;		
	
		//2. Calculation of second term
		//r_k = #parameters for estimated distributions + k - 1
		//Here, such number of parameters is 2*k since we have 2 parameters (mean and std) for every single estimated pdf
		//Also, k = Ns
		double r_k = 2*Ns + Ns - 1;
		MDL += (r_k/2) * (Math.log(points.size())/Math.log(Math.E)); //(r_k/2)*ln(N)
		
		return MDL;
	}

	public static void mainSegmentationProcess(List<Double> points,
			List<String> trip_points,
			BufferedWriter bw,
			int max_number_of_segments) throws IOException{
		//1. Calculation of Delta for all form of segments in given trajectory		
		double[][] delta = calculateDelta(points);
	
		//2. Optimization to find the best value of Ns
		double MDL = Double.MAX_VALUE;
		int bestNs = -1;
		int[][] optimizaedIndex = null;
						
		for(int Ns = 1; Ns <= max_number_of_segments; Ns++){
			//3. Finding optimal segmentation by having N_s as number of existing segments in given trajectory
			double[][] I = new double[Ns][points.size()];		
			//In addition to minimum values, we need minimum indexes which show the best breaking points for segments
			int[][] Index = new int[Ns][points.size()];
			dynamicProgramingSegmentation(points, Ns, delta, I, Index);
			double mdl = MinimumDescriptionLength(points, Index, Ns);
			if(mdl < MDL){
				MDL = mdl;
				bestNs  = Ns;
				optimizaedIndex = Index;
			}
			System.out.print(".");
		}
		
		//3. Using optimize I and Index to get the optimal segment boundaries			
		HashSet<Integer> segmentsPoints = new HashSet<Integer>();
		int lastEndPoint = points.size()-1;
		for(int k=bestNs-1; k>= 0; k--){
			segmentsPoints.add(optimizaedIndex[k][lastEndPoint]);
			lastEndPoint = optimizaedIndex[k][lastEndPoint] - 1;
		}
		
		//4. Writing output
		for(int i=0;i<trip_points.size();i++){
			if (segmentsPoints.contains(i))
				bw.write(trip_points.get(i) + "1\n"); // meaning that this point is start of a new segment
			else
				bw.write(trip_points.get(i) + "0\n"); // meaning that this point is not start of any segment
		}
	}
}

File: python/Dynamic_Programming_Segmentation.py
# ## What Does This Script Do?

# This script performs trajectory segmentation, where you can find a detailed description of segmentation process in section 3.4 of the following paper:
# * [Discovery of Driving Patterns by Trajectory Segmentation](https://arxiv.org/pdf/1804.08748.pdf)
# 
# __Input__: transformed trajectory dataset, where each record of each trajectory has the following attributes:
# * `TripId`: id of trajectory (a string)
# * `TimeStep`: time step identifier for a record of a trajectory (an integer)
# * `ProbDissimilarity`: probability dissimilarity value for the current record of a trajectory (a float). This is the signal value for a record. 
# * `Lat`: latitude coordinate of GPS (a float)
# * `Lng`: longitude coordinate of GPS (a float)
# * `Speed`: ground speed in km/h (a float)
# * `Acceleration`: acceleration in m/s^2 (a float)
# * `Heading`: change of heading with respect to previous time step in degrees (a float)
# 
# Intput data must be specified in terms of a single csv file named as `ProbabilisticDissimilarities.csv`, and the input file must be placed inside `/prerequisiteFiles` directory. 
# 
# __Outputs__: this code generates a single csv file (inside the `/output` directory) that contains trajectory segmentation results. This file has the following attributes:
# * `TripId`: id of trajectory (a string)
# * `TimeStep`: time step identifier for a record of a trajectory (an integer)
# * `Speed`: ground speed in km/h (a float)
# * `Acceleration`: acceleration in m/s^2 (a float)
# * `HeadingChange`: change of heading with respect to previous time step in degrees (a float)
# * `Lat`: latitude coordinate of GPS (a float)
# * `Lng`: longitude coordinate of GPS (a float)
# * `PMD`: probabilistic movement dissimilarity (or probability dissimilarity) value for the current record of a trajectory (a float). This is the signal value for a record. 
# * `StartOfSegment`: an indicator that specifies whether a given record is the start of a new segment or not. A value of 1 indicates that a record is start of a new segment. 

import numpy as np
import math
import time

e  = math.e
pi = math.pi
float_max_value = 1000000.0

class tripTuple:
    def __init__(self, time_step, speed, acceleration, heading, latitude, longitude):
        self.speed = speed
        self.acceleration = acceleration
        self.heading = heading
        self.lat = latitude
        self.lng = longitude
        self.time_step = time_step


# ### Calculate Delta

# Calculation of Delta for all form of segments in given trajectory
# delta_i(n_(i-1) , n_i - 1): the formula before (3) in paper! where n_(i-1) and n_i - 1 can be any possible pairs in trajectory!
# Simply, here we have a window of dynamic size which scans whole trajectory! we will find the likelihood of that windows based on... 
# ...distribution of its points. Such distribution is a normal distribution and we have the assumption of independence of nodes ...
# ... in order to make simplification in problem

def estimate_mean_std(input):
    mean = 0
    for i in input:
        mean += i
    mean /= len(input)
    
    std = 0
    for i in input:
        std += (i-mean)**2
    std = math.sqrt(std/len(input))
    
    return mean, std

def ln_of_normal_distribution(x, mean, std):
    # res = ln(N(x, mean, std)) where N stands for normal distribution. 
    # N(x, mean, std) = (1/std * (2PI)^0.5) * exp(- (x - mean)^2 / (2 * std^2) )
    try:
        res = -1 * (math.log(std * math.sqrt(2 * pi)) + (math.pow(x-mean, 2)/(2 * math.pow(std, 2))))
        return res
    except: 
        return np.nan
    
def calculateDelta(points): # vectorize calculations
    delta = np.full((len(points)-1, len(points)), float_max_value)
    
    for i in range(len(points)-1):
        for j in range(len(points)):
            if (j-i+1) >= 2: # the length of a segment should be at least 2
                mean,std = estimate_mean_std(points[i:j+1])
                # get the ln value based on points in this segment and their corresponding distribution
                value = 0
                # The goal is minimizing the delta! So, when we have standard deviation = 0, this means no changes is observing in ...
                # ...probabilistic dissimilarity values in range i to j! So, no change, no std! In other words, we have a uniform distribution...
                # ...for this range! In this way, I decided to let the delta to be 0 in such situation. 
                if std!=0:
                    mean = np.full(j+1-i, mean) # to prepare input for a vectorized  calculation
                    std  = np.full(j+1-i, std) # to prepare input for a vectorized calculation
                    v = points[i:j+1]
                    # -1 * (math.log(std * math.sqrt(2 * pi)) + (math.pow(x-mean, 2)/(2 * math.pow(std, 2))))
                    value = np.log(std * np.sqrt(2*pi)) + np.power(v-mean, 2)/(2*np.power(std, 2))
                    value = np.sum(value)
                delta[i][j] = value
    return delta
                
    
# ### Segmentation Process as a Dynamic Programming Algorithm

def dynamicProgramingSegmentation(points, Ns, delta):
    # Finding optimal segmentation by having N_s as number of existing segments in given trajectory
    # Here, we will find I_k(L) for all k = 1,2,...,N_s and for all L = 1,2,...,N
    # The approach for finding these values is based on dynamic programming formula...
    # ... which is provided as relation (4) in paper

    # initialization of I
    I = np.full((Ns, len(points)), float_max_value)
    
    # Initialization of Index
    # In addition to minimum values, we need minimum indexes which show the best breaking points for segments
    Index = np.full((Ns, len(points)), 0, dtype=int)
    
    # Now, we are going to find the optimal values
    for k in range(Ns):
        # Base case: this mean just having a single segment or I_1(L)
        # Note: here k is 0 but based on paper formulation is 1
        if k == 0:
            # Here, the assumption is L>=2; but I considered this once that delta is calculated
            for L in range(len(points)):
                I[k][L] = delta[k][L]  # Note, segments of length 0 or 1 have delta value as +infinity!
        

        # Rest of the cases (having more than one segment): Well, here is the idea of dynamic programming
        else:
            # why (k*2 + 1): look at the printed picture! I have to have some lower level for L! when I have a couple of segments, ... 
            # ...the minimum criteria is I have a segment from 0 to 1. And the 2nd segment would be from 2! So, when k=1, the min length...
            # ... for L should be 3 in order to have at least two separate segments. 
            for L in range(k*2+1, len(points)):
                min_value = float_max_value
                min_index = -1
                # I. Why (L-1)? since I want to let the last segment be at least from L-1 to L. That is, the minimum length is 2!
                # II. Look, here I have equality condition for L not just lower than! since L is index of upper loop and is precise. 
                # III. Why nk_1 = k*2? look at the last line of formula in left column of page 2 of paper. we have nk_1 as starting index ...
                # ... of last segment. When have k = 1 or just want two segments, such index is at least 2! this means k*2. 
                for nk_1 in range(k*2, L):
                    value = I[k-1][nk_1-1] + delta[nk_1][L]
                    if value < min_value:
                        min_value = value
                        min_index  = nk_1
                I[k][L]     = min_value
                Index[k][L] = min_index
                
    return I, Index
        


# ### Minimum Description Length to Find Optimal Number of Segments

def MinimumDescriptionLength(points, Index, Ns):
    # Here, the goal is to find the goodness of fit without any kind of over-fitting or a problem which have unseen data. 
    # We use the formula (7) in paper "Optimal segmentation of signals and its application to image denoising and boundary feature extraction (2004)" which is written based on a well-known research paper: "modeling by shortest data description, 1978"		
    MDL = 0

    # 1. Calculation of first term
    mle = 0

    segment_end = len(points) - 1
    segment_begin = -1
    
    for k in reversed(range(0, Ns)):
        thisSegmentMLE = 0
        segment_begin = Index[k][segment_end]
        mean,std = estimate_mean_std(points[segment_begin:segment_end+1])
        for i in range(segment_begin,segment_end):
             thisSegmentMLE += ln_of_normal_distribution(points[i], mean, std)
        mle += thisSegmentMLE
        segment_end = segment_begin-1
    
    # there is a negative sign behind the ln in equation (7)
    mle *= -1
    MDL = mle

    # 2. Calculation of second term
    # r_k = #parameters for estimated distributions + k - 1
    # Here, such number of parameters is 2*k since we have 2 parameters (mean and std) for every single estimated pdf
    # Also, k = Ns
    r_k = 2*Ns + Ns - 1
    MDL += (r_k/2) * math.log(len(points)); # (r_k/2)*ln(N)

    return MDL


# ### Segmentation Workflow

class trip_tuple:
    def __init__(self, time_step, speed, acceleration, heading, latitude, longitude, pmd):
        self.speed = float(speed)
        self.acceleration = float(acceleration)
        self.heading = float(heading)
        self.lat = float(latitude)
        self.lng = float(longitude)
        self.time_step = int(time_step)
        self.pmd = float(pmd) # probabilistic movement dissimilarity

def segmentation_process(trip_id, # trip id
                         points,  # contains probabilistic dissimilarity values
                         trip_points, # contains trip points
                         max_number_of_segments, # the maximum number of segments that we allow
                         writer # writer to print output 
                        ):
    # 1. Calculation of Delta for all form of segments in given trajectory
    print ('Building delta for {} ... '.format(trip_id), end='')
    start = time.time()
    delta = calculateDelta(points)
    print ('completed in {:.1f} sec!'.format(time.time()-start))
    
    # 2. Optimization to find the best value of Ns
    MDL = float_max_value
    bestNs = -1
    optimizedIndex = None
    print ('Segmenting trajectory {} '.format(trip_id), end=''),
    
    start = time.time()
    for Ns in range(1, max_number_of_segments+1):
        # 3. Finding optimal segmentation by having N_s as number of existing segments in given trajectory
        # In addition to minimum values, we need minimum indexes which show the best breaking points for segments
        I,Index = dynamicProgramingSegmentation(points, Ns, delta)
        mdl = MinimumDescriptionLength(points, Index, Ns)
        if mdl < MDL:
            MDL = mdl
            bestNs = Ns
            optimizedIndex = Index
        print ('.', end='')
    print (' completed in {:.1f} sec!'.format(time.time() - start))
    
    # 3. Using optimize I and Index to get the optimal segment boundaries
    segmentPoints = set()
    lastEndPoint = len(points)-1
    for k in reversed(range(bestNs)):
        segmentPoints.add(optimizedIndex[k][lastEndPoint])
        lastEndPoint = optimizedIndex[k][lastEndPoint] - 1
        
    for i in range(0, len(trip_points)):
        if i in segmentPoints:
            writer.write('{},{},{},{},{},{},{},{},{}\n'.format(trip_id, 
                                                       trip_points[i].time_step,
                                                      trip_points[i].speed,
                                                      trip_points[i].acceleration,
                                                      trip_points[i].heading,
                                                      trip_points[i].lat,
                                                      trip_points[i].lng,
                                                      trip_points[i].pmd,
                                                      1))
        else:
            writer.write('{},{},{},{},{},{},{},{},{}\n'.format(trip_id, 
                                                       trip_points[i].time_step,
                                                      trip_points[i].speed,
                                                      trip_points[i].acceleration,
                                                      trip_points[i].heading,
                                                      trip_points[i].lat,
                                                      trip_points[i].lng,
                                                      trip_points[i].pmd,
                                                      0))
            


writer = open('output/segmentation_results.csv', 'w')
writer.write('TripId,TimeStep,Speed,Acceleration,HeadingChange,Latitude,Longitude,PMD,StartOfSegment\n')
max_number_of_segments = 50
n_trajectories = 0

path = r'C:\Users\sobhan\workspace\TrajectorySegmentation'
with open('prerequisiteFiles/ProbabilisticDissimilarities.csv', 'r') as reader:
    header = True
    trip_id = ''
    points = [] # for probability dissimilarity values
    trip_points = [] # for all trip points
    
    for line in reader:
        if header:
            header = False
            continue
        parts = line.replace('\r','').replace('\n','').split(',') # TripId,TimeStep,ProbDissimilarity,Lat,Lng,Speed,Acceleration,Heading
        if parts[0] == trip_id:
            points.append(float(parts[2]))
            trip_points.append(trip_tuple(parts[1], # time step
                                         parts[5], # speed
                                         parts[6], # acceleration
                                         parts[7], # heading change
                                         parts[3], #latitude
                                         parts[4], #longitude
                                         parts[2] # pmd
                                ))
        else:
            if trip_id != '':
                # do segmentation
                segmentation_process(trip_id, points, trip_points, max_number_of_segments, writer)
                n_trajectories += 1
            trip_id = parts[0]
            points = [float(parts[2])]
            trip_points = [trip_tuple(parts[1], # time step
                                         parts[5], # speed
                                         parts[6], # acceleration
                                         parts[7], # heading change
                                         parts[3], #latitude
                                         parts[4], #longitude
                                         parts[2] # pmd
                                )
                ]
        
    
# do segmentation for the last trajectory
segmentation_process(trip_id, points, trip_points, max_number_of_segments, writer)
n_trajectories += 1
writer.close()   
print ('\nDone with segmentation of {} trajectories!'.format(n_trajectories))


# ### Notes
# 
# * Building delta is a preprocessing step prior to dynamic programming segmentation, which takes surprisingly longer than its Java implementation to run. 
# 
# * The python implmentation of segmentation takes way longer than its Java implmentation, altough both perform the same algorithm. 
